### ISSUE TITLE
platforms without sync_file_range

### ISSUE THREAD

**Issue by ThomasWaldmann** [borgbackup/borg#4905]
Moved from a comment in #4901 by @willsALMANJ:

When I try to run `borg init`, I get the error `ImportError: dlopen failed: cannot locate symbol "sync_file_range" refereced by ...borg/platform/linux.cpython-38.so`. Looking in `usr/include/fcntl.h`, I found

```
#if __ANDROID_API__ >= 26
int sync_file_range(...);
#endif
```

My phone uses Android 7 which corresponds to __ANDROID_API__ level 24, so it seems that I need something with at least Android 8 to test. I could try testing in an Android VM if I can't find a newer phone to test with and someone else doesn't test it.

By the way, I get the same error when I use the prebuilt borgbackup Termux package.

To do the test, I did `pkg install liblz4 zstd openssl libacl pkg-config clang` and then `pip install pkgconfig cython` in a virtual environment before running `python setup.py install`. That all appeared to build fine. I discovered that for borg 1.10 it is necessary to set the paths for the required shared libraries with the environment variables like `BORG_OPENSSL_PREFIX` like https://github.com/termux/termux-packages/pull/4644 does, but on master the new usage of `pkgconfig` in `setup.py` works. One other note -- `clang` was needed above for `cython`; in the version of borg packaged for Termux a released version of borg is used which does not need `cython`, so that is why `cython` and `clang` are not referenced in the Termux package's `build.sh`.

**Comment by ThomasWaldmann** [comment]
We also had a sync_file_range issue on Windows Subsystem for Linux (WSL).

Maybe the way master branch deals with that (env var) should be also offered on 1.1-maint (but keeping the automated handler for WSL), so it could be also used on old android.

**Comment by ThomasWaldmann** [comment]
@willsALMANJ how does `uname -r` look like on your android phone?

is there an easy way to check the ANDROID_API level at runtime?

**Comment by willsALMANJ** [comment]
> We also had a sync_file_range issue on Windows Subsystem for Linux (WSL).
Ah, I hadn't noticed this. So for the testing I was doing based on master, I can use `BORG_WORKAROUNDS=basesyncfile` and it should work.

I also realized that the reason I thought borg worked in Termux previously was because of [this repo](https://github.com/ravenschade/borgbackup_on_android) that had a build script for Termux. In the build script, it would test for `sync_file_range` support and [patch the code](https://github.com/ravenschade/borgbackup_on_android/blob/5a77943df0de2e82b5d45f762a3e67b0d638e34c/borg_sync_file_range.patch) if support was not found.

I am not sure if there is a sufficiently clean way to implement this but it would be nice an error related to sync_file_range could mention `BORG_WORKAROUNDS=basesyncfile`. Maybe it will be easier to find in the documentation once `BORG_WORKAROUNDS` is in a released version of borg. Even better would be to fall back to basesyncfile if sync_file_range failed, but that might be too complex to implement for the small set of cases where it matters.

> is there an easy way to check the ANDROID_API level at runtime?

`uname -r` gives `3.10.49-g415000d`. The best I have found is `getprop ro.build.version.release` which gives the Android version. I think Android version corresponds to a single API version.

**Comment by ThomasWaldmann** [comment]
Can I have the full uname -r please?

**Comment by willsALMANJ** [comment]
`uname -a`: `Linux localhost 3.10.49-g415000d #1 SMP PREEMPT Tue Feb 5 14:39:18 UTC 2019 armv7l Android`

**Comment by ThomasWaldmann** [comment]
1.1.x already checks for "Microsoft" in the kernel version. Could add a check for "Android".

It would just use the same code as on BSD in that case, which does not use sync_file_range.

In master / 1.2, that magic is removed and one needs to use the BORG_WORKAROUNDS env var.

**Comment by willsALMANJ** [comment]
Hmm, the problem on Android is a little different than the WSL problem that BORG_WORKAROUNDS was created to address. In the Android case, the error is `ImportError: dlopen failed: cannot locate symbol "sync_file_range" referenced by ...borg/platform/linux.cpython-38.so`. So the error is raised when `borg.platform.linux` is imported due to the missing symbol, before the workaround logic has a chance to work (I tested and got the same error even when setting BORG_WORKAROUNDS). I poked around a little at linux.pyx, but I am not that familiar with Cython. Is there an easy way to put in a placeholder `sync_file_range` for the case where it is not found in `fcntl.h`?

**Comment by ThomasWaldmann** [comment]
Crap. Yeah, in that case it isn't that easy. So guess we need Android 8 or a better idea. :-)

**Comment by willsALMANJ** [comment]
Well, I tested one solution that worked for me. You might not like it though. I created a new file `src/borg/platform/sync.pyx` with contents:

```
from libc.stdint cimport int64_t


cdef extern from "fcntl.h":
    int sync_file_range(int fd, int64_t offset, int64_t nbytes, unsigned int flags)
    unsigned int SYNC_FILE_RANGE_WRITE
    unsigned int SYNC_FILE_RANGE_WAIT_BEFORE
    unsigned int SYNC_FILE_RANGE_WAIT_AFTER
```

and removed the `cdef extern form fcntl.h` block from `linux.pyx`. Then just above `class SyncFile(BaseSyncFile)` I added `from .sync import sync_file_range, SYNC_FILE_RANGE_WRITE, SYNC_FILE_RANGE_WAIT_BEFORE, SYNC_FILE_RANGE_WAIT_AFTER`.

With these changes, borg never tries to import a module referencing `sync_file_range` if `BORG_WORKAROUNDS=basesyncfile` is set, so it does not hit an import error.

With this method, if desired, `from .sync import sync_file_range` could also be put in a `try:` block and `basesyncfile` could be added to `workarounds` automatically in the `except` block, so that borg could fall back to base file sync without having to set BORG_WORKAROUNDS. Maybe this is adding too much complexity though.

**Comment by ThomasWaldmann** [comment]
@willsALMANJ do you have some code that's ready for a PR against master or 1.1-maint or shall I re-implement your idea?

**Comment by willsALMANJ** [comment]
I opened a PR to show what worked for me. Feel free to use it or reimplement it if you don't like something about it.

**Comment by ThomasWaldmann** [comment]
@willsALMANJ Thanks! Merged the PR into master. Did you test that code on old android?

**Comment by willsALMANJ** [comment]
Yes, I did some simple tests with creating and extracting a repo and it seemed to work. Just now I tried testing the current master with #4965 merged in but realized that master does not have the os.link fix from https://github.com/borgbackup/borg/pull/4903. I will try testing with #4970 which has the equivalent sync_file_range fix and does have the os.link fix as well.

**Comment by ThomasWaldmann** [comment]
Likely fixed by #4965 and #4970.

Not sure what you mean by "does have the os.link fix as well"?
Isn't that still to do? Are you working on a forward port of the os.link fix?

**Comment by willsALMANJ** [comment]
What I meant was that in order to test on Android 7, both the os.link fix and the sync_file_range fix are needed. #4903 added the os.link fix to 1.1-maint. #4965 added the sync_file_range fix to master. At that point, neither master nor 1.1-maint was usable because they each had only one of the two fixes. #4970 added the sync_file_range fix to 1.1-maint, so that then 1.1-maint had both fixes.

One part that might be confusing in the above -- how did I test #4965 if master didn't have the os.link fix? Locally, I had added in the fixes from #4903 manually but I didn't commit them with #4965.

**Comment by willsALMANJ** [comment]
I had a chance to look more closely at the unit test results on my phone. Here are the results:

* 45 tests in borg.testsuite.archiver failed. All of them were related to not having permission to set selinux attributes on files. I patched restore_attrs() in archive.py to ignore EACCESS errors that involed security.selinux and these tests all passed.
* ArchiverCheckTestCase.test_verify_data_unencrypted in borg.testsuite.archiver caused Termux to crash (I tried three times and it crashed each time). I didn't try stepping through with a debugger to figure out exactly what command was causing the crash.
* 4 tests in borg.testsuite.helpers (the test_get_*_dir ones) because they try to create a directory in /var and /var is read-only.
* 4 tests in borg.testsuite.platform fail.
  - Two swidth tests fail. The results show that borg calculated the Japanese characters as being one character in size each instead of two.
  - Two acl tests failed. From what I could tell, it seemed like the tests expected results with "root" for user and group ID's, but borg was getting "nobody" instead.

I think these failures just highlight some edge cases of how user privileges work in Termux (except for the swidth ones -- maybe those are a problem if you try to restore files with unicode symbols in the names?), but that basically borg is working okay in Termux. If you think any of that should be looked at any more, let me know.

**Comment by willsALMANJ** [comment]
I revisited the development docs like you suggested in another issue and saw that the standard way to run the tests was with fakeroot. I guess the four helpers failures and thw two acl failures are from not using fakeroot. I tried using it on Termux but pytest hung in test discovery for some reason.

So it is only the two character width tests and the crashing archiver width test that I don't fully understand the implications of.

**Comment by ThomasWaldmann** [comment]
Yeah, root or fakeroot. fakeroot sometimes has its very own issues.

**Comment by willsALMANJ** [comment]
Yes, that is right. I was putting the results here to indicate how well borg worked in Termux now with this change. I think my issues with the test break down into four possible new issues:

1. Option to handle environments where security.selinux attributes can not be changed. I am not sure anything is needed here except maybe allowing the tests to pass when this attribute can not be set? Otherwise borg runs okay. It justs ends with a non-zero exit code.
2. Identify the issue with the swidth check.
3. Detect whether the tests are being run by root or within fakeroot and skip the 6 tests that need root if not.
4. Look into the bug related to test_verify_data_unencrypted causing the process to crash. This one is very strange. It crashes on the cmd() call for `borg check --verify-data`. I tried calling this from the command line on the repository left behind after the crash and it worked fine that way. I tried stepping through with the debugger. In verify_data() in archive.py, it got stuck in the `self.key.decrypt()` step. Within decrypt(), it got stuck in the `self.decompress(payload)` step. It seemed to be stuck in an infinite loop in the debugger between the get() and resize() methods of the Buffer class of helpers.py. It was bouncing back and forth between them growing the buffer size each time. I couldn't see what was causing the buffer to keep growing, but I assume if I had kept stepping through I would have seen the buffer hit some limit that caused the process to crash.
